libevmasm/ThirdOptimizerTesting.cpp

```diff
@@ -0,0 +1,279 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libevmasm/ThirdOptimizerTesting.h>
+#include "ThirdOptimizerTesting.h"
+#include <algorithm>
+#include <tuple>
+
+template<typename It>
+It backExpressionBegin(It _begin, It _end)
+{
+	AssemblyItem const& topLevelItem = *(_end - 1);
+	signed arguments = topLevelItem.arguments();
+	auto currentEnd = _end - 1 /* top-level operation */;
+	while (arguments > 0)
+	{
+		// TODO this code might break if anything returns more than 1 value, since we can't split things up
+		arguments -= static_cast<AssemblyItem const&>(*(currentEnd - 1)).returnValues();
+		currentEnd = backExpressionBegin(_begin, currentEnd);
+		assertThrow(currentEnd >= _begin, OptimizerException, "currentEnd is before begin");
+	}
+	return currentEnd;
+}
+
+vector<vector<AssemblyItem>> parseExpressions(vector<AssemblyItem> const& _items, signed _limit = -1)
```

Comment:
This looks like rather complicated code and it seems to do a lot of copying.
Do you think it might be possible for the `pattern::match` function to just forward the appropriate offset adjustments to its arguments?